src/protocol-handlers/urls.js

```diff
@@ -0,0 +1,147 @@
+const toUri = require('multiaddr-to-uri')
+const { CID } = require('multiformats/cid')
+const fetch = require('node-fetch')
+
+const DEFAULT_GATEWAY = 'https://dweb.link'
+const LOCAL_HOSTNAMES = ['127.0.0.1', '[::1]', '0.0.0.0', '[::]']
+
+/**
+ * @typedef ParsedUrl
+ * @type {object}
+ * @property {string} protocol
+ * @property {string} hostname
+ * @property {string} path
+ */
+
+/**
+ * Parses an IPFS/IPNS/dWeb URL to be handled by IPFS Desktop.
+ *
+ * @param {string} url
+ * @returns {ParsedUrl}
+ */
+function parseUrl (url) {
+  let protocol = null
+  let hostname = null
+  let path = '/'
+
+  if (url.startsWith('ipfs://')) {
+    protocol = 'ipfs'
+    hostname = url.slice(7)
+  } else if (url.startsWith('ipns://')) {
+    protocol = 'ipns'
+    hostname = url.slice(7)
+  } else if (url.startsWith('dweb:/ipfs/')) {
+    protocol = 'ipfs'
+    hostname = url.slice(11)
+  } else if (url.startsWith('dweb:/ipns/')) {
+    protocol = 'ipns'
+    hostname = url.slice(11)
+  } else {
+    return null
+  }
+
+  if (hostname.includes('/')) {
+    const [first, ...rest] = hostname.split('/')
+    hostname = first
+    path = '/' + rest.join('/')
+  }
+
+  return { protocol, hostname, path }
```

Comment:
using the `URL` class should help here.

We should add some tests around this method to ensure we're parsing things properly.

Reply:
I think I used (or tried) before. But I got something along these lines, which seems much more complicated than the current approach:

```js
function parseUrl (urlStr) {
  let protocol = null
  let hostname = null
  let path = '/'

  try {
    const url = new URL(urlStr)

    if (['ipfs:', 'ipns:'].includes(url.protocol)) {
      protocol = url.protocol.slice(4)
      hostname = url.pathname
    } else if (url.protocol === 'dweb:') {
      if (url.pathname.startsWith('/ipfs/')) {
        protocol = 'ipfs'
        hostname = url.pathname.slice(6)
      } else if (url.pathname.startsWith('/ipns/')) {
        protocol = 'ipns'
        hostname = url.pathname.slice(6)
      } else {
        return null
      }
    } else {
      return null
    }
  } catch (_) {
    return null
  }

  if (hostname.startsWith('/')) {
    hostname = hostname.substring(1)
  }

  if (hostname.includes('/')) {
    const [first, ...rest] = hostname.split('/')
    hostname = first
    path = '/' + rest.join('/')
  }

  return { protocol, hostname, path }
}
```